packages/sandbox/src/helpers/subscriptionPostMessageRelayHelpers.ts

```diff
@@ -129,26 +178,77 @@ class SubscriptionClient {
     }
     assertUnreachable(this.protocol);
   }
+  onStopListening() {
+    if (this.protocol === 'http-multipart') {
+      this.multipartClient = {
+        connected: false,
+        onConnectedCallback: this.multipartClient?.onConnectedCallback,
+        stopListeningCallback: this.multipartClient?.stopListeningCallback,
+      };
+      this.multipartClient?.stopListeningCallback?.();
+      return;
+    }
+    if (this.protocol === 'graphql-ws') {
+      return this.graphWsClient.dispose();
+    }
+    if (this.protocol === 'subscriptions-transport-ws') {
+      return this.transportSubscriptionClient.close();
+    }
+    assertUnreachable(this.protocol);
+  }
 
   request(
     params: OperationOptions & {
       query: string;
-      variables: JSONObject;
+      variables: Record<string, string> | undefined;
       operationName: string | undefined;
+      httpMultipartParams?: HTTPMultipartParams;
+      embeddedIFrameElement: HTMLIFrameElement;
+      embedUrl: string;
+      operationId: string;
     }
   ) {
     return {
-      subscribe: (
+      subscribe: async (
         subscribeParams: Observer<ExecutionResult<ObjMap<unknown> | JSONValue>>
       ) => {
+        if (this.protocol === 'http-multipart' && params.httpMultipartParams) {
+          // TODO(Maya): right now the way the execute / repond is set up is confusing for
+          // subscriptions with this addition. We don't call `subscribeParams` in this
+          // branch, we respond in `executeOperation` with pm's to the embed.
+          // I would like to clean this up in both the explorer & sandbox, but first I would
+          // like to move the shared code (subscription code, request code) into a shared npm package
```

Comment:
This whole repo needs some cleaning up because 
1. we need to keep types up to date with the studio-ui repo
2. we need to maintain the same code in sandbox & explorer packages

I'd like to spend some time creating a shared package for sandbox & explorer instead of copying all of these changes into the explorer code, this should mitigate #2. I'd also like to go through and audit the types and match them closer to studio-ui

Reply:
Asked chang about sharing the code in a shared npm package, and we decided to do it after file upload. So I will put up another PR supporting multi part subs in the explorer package